modules/gatsby/src/gatsby-node/create-example-pages.js

```diff
@@ -0,0 +1,160 @@
+const path = require('path');
+const { log, COLOR } = require('../utils/log');
+
+const EXAMPLES_PAGE = path.resolve(__dirname, '../templates/examples.jsx');
+const EXAMPLE_PAGE = path.resolve(__dirname, '../templates/example-n.jsx');
+
+function getExampleThumbnails({ allFile, allImageSharp }) {
+  // this function associates the path of an original image
+  // with the public url of a thumbnail, resized server side by imageSharp.
+  if (
+    !allFile ||
+    !allFile.edges ||
+    !allFile.edges.length ||
+    !allImageSharp ||
+    !allImageSharp.edges ||
+    !allImageSharp.edges.length
+  ) {
+    log.log({ color: COLOR.YELLOW, priority: 1 }, `No thumbnails created.`)();
+    return {};
+  }
+  // in a first pass, we create a lookup - original path to internal gatsby node id.
+  /* eslint-disable no-param-reassign */
+  const idLookup = allFile.edges.reduce((lookup, { node }) => {
+    lookup[node.id] = node.relativePath;
+    return lookup;
+  }, {});
+
+  // in a second path, we associate each thumbnail with the path of the original
+  // image, using the node id of this original image and the lookup
+  // we just created.
+
+  const pathLookup = allImageSharp.edges.reduce((lookup, { node }) => {
+    const originalImageId = node.parent.id;
+    const originalImagePath = idLookup[originalImageId];
+    lookup[originalImagePath] = node.resize.src;
+    return lookup;
+  }, {});
+
+  /* eslint-enable no-param-reassign */
+  return pathLookup;
+}
+
+function createExamplePages(EXAMPLES, thumbnailsPublicUrls = {}, createPage) {
+  if (EXAMPLES.length === 0 || EXAMPLES[0].title === 'none') {
+    return;
+  }
+  // matches public urls to paths of images
+  const examplesWithImage = EXAMPLES.map(example => ({
+    ...example,
+    imageSrc: thumbnailsPublicUrls[example.image]
+  }));
+
+  createPage({
+    component: EXAMPLES_PAGE,
+    path: '/examples',
+    context: {
+      toc: 'examples',
+      examples: examplesWithImage
+    }
+  });
+
+  EXAMPLES.forEach(example => {
+    const exampleName = example.title;
+
+    log.log(
+      { color: COLOR.CYAN, priority: 1 },
+      `Creating example page ${JSON.stringify(example)}`
+    )();
+
+    createPage({
+      path: example.path,
+      component: EXAMPLE_PAGE,
+      context: {
+        slug: exampleName,
+        toc: 'examples'
+      }
+    });
+  });
+}
+
+module.exports = function prepareExamplePages({ graphql, actions }) {
```

Comment:
Nit: Maybe dedicate this long function just to doing the query, rather than hiding an example creation stanza at the very end?

Reply:
i'm not sure what's a better form. could you take a stab? 
the point of this function is to do 

query w/ images -> if success, create thumbnails, then create examples with images
|
v 
if failure, query w/o images -> if success, create examples w/o images